### PR TITLE
Unveil type check part in _VerifyGeneratedGradients that was not actu…

### PR DESCRIPTION
**Pull request by YuMS**
…ally used

### CODE REVIEW

File: tensorflow/python/ops/gradients_impl.py [hunk 1]
```diff
@@ -273,28 +273,28 @@ def _VerifyGeneratedGradients(grads, op):
...
-          raise TypeError("Gradient type %s generated for complex-valued op %s"
-                           " with type %s must be complex" %
-                           (dtypes.as_dtype(grad.dtype).name, op.node_def,
-                            dtypes.as_dtype(inp.dtype).name))
-      else:
-        raise TypeError("Gradient type %s generated for op %s "
-                         "with type %s must be either real or complex" %
-                         (dtypes.as_dtype(grad.dtype).name, op.node_def,
-                          dtypes.as_dtype(inp.dtype).name))
+  for i in xrange(len(grads)):
```
Comment (vrv) [inline tensorflow/python/ops/gradients_impl.py]: Whoa, this was bad, thank you for finding this.

The fact that no tests failed means that there's no test that exercises this.  Can you conceive of a test in the gradients_test.py file that would have exercised this code?  Thank you!

**Comment by vrv** [comment]
cc @yaroslavvb in case he has an idea of how he wanted this code to be checked in the first place

**Comment by yaroslavvb** [comment]
rats! Python indendation!

I added the check because @girving asked me to add a check that gradients must be complex when inputs are complex in https://github.com/tensorflow/tensorflow/pull/6868

But, I'm not sure of a way to exercise complex gradients, @girving any idea? Since nobody complained, maybe we can just drop that check completely?

The original reason for PR is that it enforced that activation value and backprop value have the same type, but we wanted to mix types (float32, float16). What if someone decides to mix real/complex types in the future?

File: tensorflow/python/ops/gradients_impl.py (hunk 1 above)
Comment (YuMS) [inline tensorflow/python/ops/gradients_impl.py]: I'd like to, when your final call is to keep these line.

**Comment by girving** [comment]
It's hard for me to imagine a use of replacing real with complex or vice versa during backprop.  You'd be throwing away half the information, or adding zeros for some reason.

Exercising complex gradients is pretty easy: just make a complex constant and square it, for instance.

**Comment by yaroslavvb** [comment]
I think something extending "synthetic-gradients" idea could mix complex real for gradient -- they use a different function for forward and backward, so you throw away half the information, but gain some regularization/efficiency.

I don't have any strong opinion for what to do here, I think the number of people doing complex gradients is pretty small

**Comment by vrv** [comment]
Let's just get rid of the check -- we can add it back later, it's clear no one was depending on it.

**Comment by yaroslavvb** [comment]
@YuMS could you just delete that whole block? Thanks for the find

**Comment by vrv** [comment]
@yaroslavvb GitHub now allows you to make edits like these in their PR (click the pencil button in the review tab) -- if you think this is the way to go.

**Comment by yaroslavvb** [comment]
Yes, that's super convenient, although last time I tried it, it broke the CLA bot, lets see what happens this time

**Comment by yaroslavvb** [comment]
OK, looks like CLA bot is still confused by github inline edits, same as in https://github.com/tensorflow/tensorflow/pull/7442#issuecomment-279272947 , it needs someone to manually add "cla: yes" label @martinwicke

**Comment by vrv** [comment]
It's possible that it doesn't like it when the person making the edit isn't an admin (it allows me to do it all the time without complaining...).  IN any case, this is fine, so I'll test and merge.

@tensorflow-jenkins test this please

**Comment by martinwicke** [comment]
It just doesn't like it period. It's fine to ignore the CLAbot if it
complains about the consent thing and all the commits are from the
submitter or maintainers.

**Comment by vrv** [comment]
I've been making edits all week and it never complained :)
